python/ray/_raylet.pyx

```diff
@@ -284,145 +282,77 @@ class StreamingObjectRefGenerator:
         Args:
             timeout_s: If the next object is not ready within
                 this timeout, it returns the nil object ref.
-            sleep_interval_s: busy waiting interval.
-            unexpected_network_failure_timeout_s: If the
-                task is finished, but the next ref is not
-                available within this time, it will hard fail
-                the generator.
         """
-        obj = self._handle_next_sync()
-        last_time = time.time()
-
-        # The generator ref will be None if the task succeeds.
-        # It will contain an exception if the task fails by
-        # a system error.
-        while obj.is_nil():
-            error_ref = self._handle_error(
-                False,
-                last_time,
-                timeout_s,
-                unexpected_network_failure_timeout_s)
-            if error_ref is not None:
-                return error_ref
-
-            time.sleep(sleep_interval_s)
-            obj = self._handle_next_sync()
-
-        return obj
-
-    async def _next_async(
-            self,
-            timeout_s: float = -1,
-            sleep_interval_s: float = 0.0001,
-            unexpected_network_failure_timeout_s: float = 30):
-        """Same API as _next_sync, but it is for async context."""
-        obj = await self._handle_next_async()
-        last_time = time.time()
-
-        # The generator ref will be None if the task succeeds.
-        # It will contain an exception if the task fails by
-        # a system error.
-        while obj.is_nil():
-            error_ref = self._handle_error(
-                True,
-                last_time,
-                timeout_s,
-                unexpected_network_failure_timeout_s)
-            if error_ref is not None:
-                return error_ref
-
-            await asyncio.sleep(sleep_interval_s)
-            obj = await self._handle_next_async()
-
-        return obj
-
-    async def _handle_next_async(self):
-        try:
-            return self._handle_next()
-        except ObjectRefStreamEneOfStreamError:
-            raise StopAsyncIteration
+        if not hasattr(self.worker, "core_worker"):
+            raise ValueError(
+                "Cannot access the core worker. "
+                "Did you already shutdown Ray via ray.shutdown()?")
+        core_worker = self.worker.core_worker
+
+        # Wait for the next ObjectRef to become ready.
+        ref = core_worker.peek_object_ref_stream(
+            self._generator_ref)
+        ready, unready = ray.wait(
+            [ref], timeout=timeout_s, fetch_local=False)
+        if len(unready) > 0:
+            return ObjectRef.nil()
 
-    def _handle_next_sync(self):
         try:
-            return self._handle_next()
-        except ObjectRefStreamEneOfStreamError:
-            raise StopIteration
+            ref = core_worker.try_read_next_object_ref_stream(
```

Comment:
Why do we reset the ref, don't we already have the ref from peek_object_ref_stream?

Reply:
Oh, you just use it to see if this is the last one. Can we add a check to make sure the previous ref is the same as this one.